style(mypy): enforcing mypy typing for superset.connectors module

john-bodley opened the pull request:
### SUMMARY

Adding `mypy` type enforcement to the `superset.connectors` module.

Note this PR was much more involved/complex than I had originally thought given the dependency on other modules which were untyped. Additionally the overall code quality is quite poor; large complex functions, numerous variables which were either optional or with mixed types (`Union`), variable redefinition, etc. meant that the logic was quite hard to grok.

Note this PR may introduce some regressions due to the necessary restructuring, however I sense long-term benefits of having the entire repo typed outweighs the short-term concerns regarding possibly (minor) regressions.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TEST PLAN

CI.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Commit:
style: enforcing mypy typing for connectors

villebro commented:
Oh boy, gonna need to take some sedatives before this 😄 Thanks for taking this on @john-bodley , I'll review shortly.

john-bodley commented:
@villebro I started out thinking this would be a somewhat mindless/trivial change to kill some time whilst sheltering-in-place. It definitely became more complex/tedious than I initially had hoped.

john-bodley commented on superset/connectors/sqla/models.py:
```diff
@@ -841,6 +858,7 @@ def get_sqla_query(  # sqla
 
         where_clause_and = []
         having_clause_and: List = []
+        assert filter
```
Note this could be `None` per the function signature and thus line #862 would throw an exception, however to the best of my knowledge this hasn't happened and thus `filter` is probably _never_ `None`.

john-bodley commented on superset/views/base.py:
```diff
@@ -359,19 +361,21 @@ def yaml_export(self, items):
                 record primary key to delete
         """
-        item = self.datamodel.get(primary_key, self._base_filters)
+        item = self.datamodel.get(primary_key, self._base_filters)  # type: ignore
```
Because FAB isn't typed and there's no stubs it seems simpler to just ignore these.

john-bodley commented on superset/viz.py:
```diff
@@ -1055,9 +1055,9 @@ def query_obj(self):
-        self.x_metric = form_data.get("x")
-        self.y_metric = form_data.get("y")
-        self.z_metric = form_data.get("size")
+        self.x_metric = form_data["x"]
```
I perceive these are all required otherwise `utils.get_metric_name(...)` etc. would return `None` and cause issues when trying to index the Pandas DataFrame.

villebro commented on superset/connectors/druid/models.py:
```diff
@@ -449,7 +460,7 @@ def get_perm(self) -> Optional[str]:
     @classmethod
     def import_obj(cls, i_metric: "DruidMetric") -> "DruidMetric":
-        def lookup_obj(lookup_metric: DruidMetric) -> Optional[DruidMetric]:
+        def lookup_obj(lookup_metric: "DruidMetric") -> Optional["DruidMetric"]:
```
Not sure how a nested function referencing its parent method's class is handled by mypy, but if this was working before, we should probably be ok leaving this without quotes?

villebro commented on superset/typing.py:
```diff
@@ -16,15 +16,31 @@
 FilterValues = Union[FilterValue, List[FilterValue], Tuple[FilterValue]]
+Granularity = Union[str, Dict[str, Union[str, float]]]
+Metric = Union[Dict[str, str], str]
+QueryObject = Dict[str, Any]
```
We should probably call this `QueryObjectDict` to distinguish from `superset.common.query_object.QueryObject`.

villebro commented on superset/connectors/druid/models.py:
```diff
@@ -558,8 +569,8 @@ def get_perm(self) -> str:
         )
 
-    def update_from_object(self, obj):
-        return NotImplementedError()
```
😄

villebro commented on superset/connectors/druid/models.py:
```diff
@@ -817,7 +829,7 @@ def granularity(
         }
 
-        granularity: Dict[str, Union[str, float]] = {"type": "period"}
+        granularity = {"type": "period"}
```
I would have thought this turns into a `Dict[str, str]` unless being explicitly being defined as `Granularity`, and thus causing problems below when numeric values are added?

john-bodley replied:
@villebro I guess it differs from the `Granularity` type which I defined previously (which adds to the confusion). When including the typing here it causes a conflict with the return type as it can also be a string per line 812. 

A couple of learnings from having typed a bunch of Python code: 

1. `Optional[...]` values are difficult to handle and thus should be avoided if possible. It seems they're often used by laziness. 
2. Mixed types, i.e., those requiring a `Union[...]` normally indicates poor quality and unnecessarily complex code. 

I think the return type of `granularity` is a clear indication of (2). Also I believe the true return type should be `Union[Dict[str, Union[str, float]], str]` but `mypy` complains on this.

villebro commented on superset/connectors/druid/models.py:
```diff
@@ -1240,7 +1268,7 @@ def run_query(  # druid
             client.scan(**qry)
         elif (IS_SIP_38 and columns) or (
-            not IS_SIP_38 and len(groupby) == 0 and not having_filters
+            not IS_SIP_38 and (not groupby or len(groupby) == 0) and not having_filters
```
I wonder if a simple falsy check here would be more pythonic: `not groupby`? I can't think of an edge case for which it wouldn't give the correct results.

john-bodley replied:
@villebro thanks. I agree and made this change elsewhere.

villebro commented on superset/connectors/druid/models.py:
```diff
@@ -1322,13 +1347,13 @@ def run_query(  # druid
+                pre_qry["dimensions"] = non_dict_dims + dict_dims  # type: ignore
 
-                order_by = None
+                order_by = None  # type: ignore
```
If we define this as `order_by: Optional[Metric]`, we can probably omit the ignore?

john-bodley replied:
@villebro the issue here is that `order_by` has already been defined on line 1291, 

```python
order_by = utils.get_metric_name(timeseries_limit_metric)
```

I declared the type as `order_by: Optional[str] = None`.

villebro reviewed (commented):
First pass

john-bodley commented:
@villebro thanks for the feedback. I've addressed your comments. Note https://github.com/apache/incubator-superset/pull/9833 is a prerequisite to fix the docs check which is failing.

john-bodley commented on superset/connectors/base/models.py:
```diff
@@ -182,7 +182,7 @@ def short_data(self) -> Dict[str, Any]:
 
     @property
-    def select_star(self):
+    def select_star(self) -> str:
         pass
```
@villebro a `NotImplementedError` causes havoc for Druid hence why this is `pass`. I made this an `Optional[str]`.

john-bodley commented on superset/utils/core.py:
```diff
@@ -985,11 +988,11 @@ def is_adhoc_metric(metric) -> bool:
 
-def get_metric_name(metric):
-    return metric["label"] if is_adhoc_metric(metric) else metric
+def get_metric_name(metric: Metric) -> str:
```
@villebro the code works if the `metric` is `None` even though the `Metric` type is explicitly `Union[Dict[str, str], str]`. Generally I think we should strive to ensure only valid inputs are provided to functions (otherwise the logic can get quite complex), i.e., I prefer, 

```python
if metric:
    name = get_metric_name(metric)
    ...
```

rather than, 

```python
name = get_metric_name(metric)

if name:
   ...
```

villebro replied:
I agree.

john-bodley commented:
@villebro would you mind taking another pass at this PR? All the CI tasks are now passing.

villebro commented:
Sure thing,@john-bodley , I'll review during the weekend

villebro commented on superset/typing.py:
```diff
@@ -16,15 +16,31 @@
+Base = Union[bytes, str]
+Status = Union[int, str]
+Headers = Dict[str, Any]
+FlaskResponse = Union[
+    flask.wrappers.Response,
+    werkzeug.wrappers.Response,
+    Base,
+    Union[Base, Status],
+    Union[Base, Status, Headers],
+]
```
As we already have a wrapping `Union` here, couldn't this be expressed as `Union[..., Base, Status, Headers]`?

john-bodley replied:
@villebro, that's actually a typo as the later `Union` should be a `Tuple` and given that `flask.wrappers.Response` is derived from `werkzeug.wrappers.Response` per [here](https://github.com/pallets/flask/blob/master/src/flask/wrappers.py#L93) this could be:

```python
FlaskResponse = Union[
    werkzeug.wrappers.Response,
    Base,
    Tuple[Base, Status],
    Tuple[Base, Status, Headers],
]
```

villebro replied:
Ok, that makes sense. Unrelated comment, it's really strange how practically none of the major libraries have implemented type annotations yet. Yes, stubs exist for many, but that doesn't really help when reading code.

villebro reviewed (approved):
One non blocking comment/question, apart from that looking very good.

villebro commented:
LGTM after the `Union` to `Tuple` change 👍

serenajiang commented on superset/utils/core.py:
```diff
@@ -256,8 +259,6 @@ def parse_human_datetime(s: Optional[str]) -> Optional[datetime]:
     >>> year_ago_1 == year_ago_2
     True
     """
-    if not s:
```
FOUND IT. @john-bodley
